Description: The tests are designed to be run after the build and before
 installation.  To accomplish this, they are designed to look for the binary
 with a hard path starting at / and ending in src/ inside the build directory.
 This patch simply calls the name of the binary (bitcoind, bitcoin-cli, etc.),
 which can be found in /usr/bin when the binary package is installed.  It also
 logs the test output to stdout instead of stderr.  Actual errors are still
 detected as such by autopkgtest.
Author: Soren Stoutner <soren@debian.org>
        Léo Haf <leo@haf.ovh>
Forwarded: not-needed
Last-Update: 2025-09-09

--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -241,11 +241,7 @@
             "bitcoin-wallet": ("bitcoinwallet", "BITCOINWALLET"),
         }
         for binary, [attribute_name, env_variable_name] in binaries.items():
-            default_filename = os.path.join(
-                self.config["environment"]["BUILDDIR"],
-                "bin",
-                binary + self.config["environment"]["EXEEXT"],
-            )
+            default_filename = binary
             setattr(self.options, attribute_name, os.getenv(env_variable_name, default=default_filename))
 
     def setup(self):
--- a/test/functional/test_runner.py
+++ b/test/functional/test_runner.py
@@ -491,7 +491,8 @@
 
     # Set up logging
     logging_level = logging.INFO if args.quiet else logging.DEBUG
-    logging.basicConfig(format='%(message)s', level=logging_level)
+    # Log to stdout instead of stderr.  Otherwise, the successful test output is interpreted as an error by autopkgtest.
+    logging.basicConfig(stream=sys.stdout, format='%(message)s', level=logging_level)
 
     # Create base test directory
     tmpdir = "%s/test_runner_₿_🏃_%s" % (args.tmpdirprefix, datetime.datetime.now().strftime("%Y%m%d_%H%M%S"))
--- a/test/util/test_runner.py
+++ b/test/util/test_runner.py
@@ -29,8 +29,10 @@
 
     parser = argparse.ArgumentParser(description=__doc__)
     parser.add_argument('-v', '--verbose', action='store_true')
+    parser.add_argument('-a', '--autopkgtest', action='store_true')
     args = parser.parse_args()
     verbose = args.verbose
+    autopkgtest = args.autopkgtest
 
     if verbose:
         level = logging.DEBUG
@@ -38,11 +40,12 @@
         level = logging.WARNING
     formatter = '%(asctime)s - %(levelname)s - %(message)s'
     # Add the format/level to the logger
-    logging.basicConfig(format=formatter, level=level)
+    # Log to stdout instead of stderr.  Otherwise, the successful test output is interpreted as an error by autopkgtest.
+    logging.basicConfig(stream=sys.stdout, format=formatter, level=level)
 
-    bctester(os.path.join(env_conf["SRCDIR"], "test", "util", "data"), "bitcoin-util-test.json", env_conf, config['components'])
+    bctester(os.path.join(env_conf["SRCDIR"], "test", "util", "data"), "bitcoin-util-test.json", env_conf, config['components'], autopkgtest)
 
-def bctester(testDir, input_basename, buildenv, component_conf):
+def bctester(testDir, input_basename, buildenv, component_conf, autopkgtest):
     """ Loads and parses the input file, runs all tests and reports results"""
     input_filename = os.path.join(testDir, input_basename)
     with open(input_filename, encoding="utf8") as f:
@@ -62,7 +65,7 @@
             continue
 
         try:
-            bctest(testDir, testObj, buildenv)
+            bctest(testDir, testObj, buildenv, autopkgtest)
             logging.info("PASSED: " + testObj["description"])
         except Exception:
             logging.info("FAILED: " + testObj["description"])
@@ -78,7 +81,7 @@
             logging.warning(f'{len(skipped_testcases)} tests skipped because {skipped_testcase_deps} is not built')
         sys.exit(0)
 
-def bctest(testDir, testObj, buildenv):
+def bctest(testDir, testObj, buildenv, autopkgtest):
     """Runs a single test, comparing output and RC to expected output and RC.
 
     Raises an error if input can't be read, executable fails, or output/RC
@@ -87,9 +90,17 @@
     # Get the exec names and arguments
     execprog = os.path.join(buildenv["BUILDDIR"], "bin", testObj["exec"] + buildenv["EXEEXT"])
     if testObj["exec"] == "./bitcoin-util":
-        execprog = os.getenv("BITCOINUTIL", default=execprog)
+        if autopkgtest:
+            # In autopkgtest, the binary is installed system-wide
+            execprog = "bitcoin-util"
+        else:
+            execprog = os.getenv("BITCOINUTIL", default=execprog)
     elif testObj["exec"] == "./bitcoin-tx":
-        execprog = os.getenv("BITCOINTX", default=execprog)
+        if autopkgtest:
+            # In autopkgtest, the binary is installed system-wide
+            execprog = "bitcoin-tx"
+        else:
+            execprog = os.getenv("BITCOINTX", default=execprog)
 
     execargs = testObj['args']
     execrun = [execprog] + execargs
